### PR TITLE
release 0.4.0

### PR DESCRIPTION
**Pull request by camshaft**

### CODE REVIEW

File: bolero-engine/src/rng.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
+use crate::{panic, Engine, SliceTestInput, TargetLocation, Test};
+use bolero_generator::driver::DriverMode;
+use core::{fmt::Debug, mem::replace};
+use rand::{rngs::StdRng, Rng, RngCore, SeedableRng};
+
+#[derive(Copy, Clone, PartialEq)]
+pub struct RngEngine {
+    pub iterations: usize,
+    pub max_len: usize,
+    pub seed: u64,
```
Comment (davidbarsky) [inline bolero-engine/src/rng.rs]: Small nit; not a blocker: I wonder if `seed` should be set like proptest's [`TestRng`](https://docs.rs/proptest/0.9.4/proptest/test_runner/struct.TestRng.html)? e.g., the `thread_rng()` could be replaced with a generic parameter implementing `https://docs.rs/rand/0.7.2/rand/trait.Rng.html` with [ThreadRng](https://docs.rs/rand/0.7.2/rand/rngs/struct.ThreadRng.html) as a default.

(I'd be happy to a PR for this. It'll allow for the engine to be created once, re-used, and replayed.)

File: bolero-model/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    fn gen_arguments() -> Self::ArgumentsGen;
+    fn gen_operation() -> Self::OperationGen;
+
+    fn init(arguments: Self::Arguments);
+    fn dispatch(&mut self, operation: Self::Operation);
+}
+
+// fn main() {
+//     fuzz!().model::<Model>();
+// }
```
Comment (davidbarsky) [inline bolero-model/src/lib.rs]: Remove?

Comment (camshaft) [reply bolero-model/src/lib.rs]: That's just an idea of what the interface would be like. The whole bolero-model crate is WIP so I'll just skip pushing it for now.

Comment (davidbarsky) [reply bolero-model/src/lib.rs]: I like this interface a _lot_ though.

File: bolero-engine/src/rng.rs (hunk 1 above)
Comment (camshaft) [reply bolero-engine/src/rng.rs]: I think we'll just stick with this for now. Feel free to open a PR!

**Pull request merged**